### DISCUSSION TITLE
Trojan Source: Stretched string

### DISCUSSION THREAD

**Discussion by danieldaeschle** [vlang/v#12697] [Ideas]
**What did you do?**
```v
fn main() {
	mut access_level := "user"
	if access_level != "user‮ ⁦// Check if admin⁩ ⁦" {
		println("You are an admin.")
	}
}
```

**What did you expect to see?**
It shouldn't output `You are an admin.`
 
**What did you see instead?**
`You are an admin.`

https://trojansource.codes/trojan-source.pdf

**Comment by medvednikov** [comment]
This is a unicode thing. Is it handled in other languages like C++ and Go?

**Comment by JalonSolov** [comment]
No.  From the PDF linked:  "We provide working examples of Trojan-Source vulnerabilities in C, C++, C#, JavaScript, Java, Rust, Go, and Python."

So even the mighty Rust, with all the security features it has, is vulnerable to this.

**Comment by bstnbuck** [comment]
The commenting-out attack scenario works under V too. However, V warns about the unused variable `is_admin`.
```v
fn main() {
    is_admin := false
    /*‮ } ⁦if is_admin⁩ ⁦ begin admins only */
        println("You are an admin.")
    /* end admins only ‮ { ⁦*/
}
```


Btw, GitHub and also VS Code already warns about the early-return, commenting-out and stretched-string attacks. The kind with homoglyphs and invisible functions not yet.
Possibly the V compiler could warn about such control characters?

**Comment by JalonSolov** [comment]
V should definitely do something, if possible.  Perhaps it can be the first language that does.

**Comment by medvednikov** [comment]
All right, let's do this in the scanner. Refuse to compile if suspicious characters are found.

**Comment by JalonSolov** [comment]
I could see allowing it in a string, but nowhere else.  Even in a string is... iffy.

**Comment by larpon** [comment]
Is there a list of all the ~~escape~~ characters 🤔?

**Comment by JalonSolov** [comment]
This is from the linked PDF:
```
TABLE I
UNICODE DIRECTIONALITY FORMATTING CHARACTERS RELEVANT TO REORDERING ATTACKS.
SEE BIDI SPECIFICATION FOR COMPLETE LIST [3].

Abbreviation Code Point   Name                         Description
LRE          U+202A       Left-to-Right Embedding      Try treating following text as left-to-right.
RLE          U+202B       Right-to-Left Embedding      Try treating following text as right-to-left.
LRO          U+202D       Left-to-Right Override       Force treating following text as left-to-right.
RLO          U+202E       Right-to-Left Override       Force treating following text as right-to-left.
LRI          U+2066       Left-to-Right Isolate        Force treating following text as left-to-right without affecting adjacent text.
RLI          U+2067       Right-to-Left Isolate        Force treating following text as right-to-left without affecting adjacent text.
FSI          U+2068       First Strong Isolate         Force treating following text in direction indicated by the next character.
PDF          U+202C       Pop Directional Formatting   Terminate nearest LRE, RLE, LRO, or RLO.
PDI          U+2069       Pop Directional Isolate      Terminate nearest LRI or RLI.
```

**Comment by bstnbuck** [comment]
> 
> 
> This is from the linked PDF:
> 
> ```
> TABLE I
> UNICODE DIRECTIONALITY FORMATTING CHARACTERS RELEVANT TO REORDERING ATTACKS.
> SEE BIDI SPECIFICATION FOR COMPLETE LIST [3].
> 
> Abbreviation Code Point   Name                         Description
> LRE          U+202A       Left-to-Right Embedding      Try treating following text as left-to-right.
> RLE          U+202B       Right-to-Left Embedding      Try treating following text as right-to-left.
> LRO          U+202D       Left-to-Right Override       Force treating following text as left-to-right.
> RLO          U+202E       Right-to-Left Override       Force treating following text as right-to-left.
> LRI          U+2066       Left-to-Right Isolate        Force treating following text as left-to-right without affecting adjacent text.
> RLI          U+2067       Right-to-Left Isolate        Force treating following text as right-to-left without affecting adjacent text.
> FSI          U+2068       First Strong Isolate         Force treating following text in direction indicated by the next character.
> PDF          U+202C       Pop Directional Formatting   Terminate nearest LRE, RLE, LRO, or RLO.
> PDI          U+2069       Pop Directional Isolate      Terminate nearest LRI or RLI.
> ```
The entire list can be found in the source referenced in the PDF: https://www.unicode.org/reports/tr9/tr9-42.html

**Comment by bstnbuck** [comment]
> 
> 
> V should definitely do something, if possible. Perhaps it can be the first language that does.

Rust has already released a patch for this... One of the first;)
https://blog.rust-lang.org/2021/11/01/Rust-1.56.1.html
https://blog.rust-lang.org/2021/11/01/cve-2021-42574.html

**Comment by JalonSolov** [comment]
Well, we can probably still beat some of the others.  :-)

**Comment by spytheman** [comment]
The others got notified in advance... From the second link:
```
Timeline of events

    2021-07-25: we received the report and started working on a fix.
    2021-09-14: the date for the embargo lift (2021-11-01) is communicated to us.
    2021-10-17: performed an analysis of all the source code ever published to crates.io to check for the presence of this attack.
    2021-11-01: embargo lifts, the vulnerability is disclosed and Rust 1.56.1 is released.
```
